### PR TITLE
Feature/health api

### PR DESCRIPTION
**Pull request by rasmusebbesen**
Adds health api with checks for redis and mysql connections

### CODE REVIEW

File: Sources/App/Utilities/Health.swift [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        }
+    }
+
+    // Health API response type
+    struct HealthResponse: Content {
+        let status: String
+        let checks: [String: [Indicator]]
+        let notes: [String]
+        let version: Int = 1
+        let releaseId: String = "1.0.0"
```
Comment (rasmusebbesen) [inline Sources/App/Utilities/Health.swift]: Perhaps these should be mutable and or initialised in the construct?

File: Sources/App/Utilities/Health.swift [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    static var componentName: String { get }
+
+    // A system component type eg. datastore, system, etc.
+    static var componentType: String { get }
+
+    // An optional descriptive name of the type of masurement made eg. connection or uptime
+    static var measurementName: String? { get }
+
+    // The function that does the heath check on a given service
+    static func isHealthy(on request: Request) -> EventLoopFuture<Health.Indicator>
```
Comment (rasmusebbesen) [inline Sources/App/Utilities/Health.swift]: Perhaps we should rename this? the "is" prefix leads to think it returns a boolean

**Review by siemensikkema — COMMENTED** [review]
Looks really good! Without making it a real package with its own repository we could still extract it into its own target and make the `App` target depend on it. Right now we have some very generic names in the global namespace (Health, System) and date formatters and such. By extracting them into a package we can scope them under "Health" (and make the date formatters internal). Remember to open up any public API by marking it public.
After that I think we're ready to merge.

File: Package.swift [hunk 3]
```diff
@@ -37,7 +40,8 @@ let package = Package(
             "Storage",
             "Submissions",
             "Sugar",
-            "Vapor"
+            "Vapor",
+            "HealthCheck"
```
Comment (siemensikkema) [inline Package.swift]: please move this up to keep it alphabetized

File: Sources/App/Setup/configure.swift [hunk 4]
```diff
@@ -5,6 +5,7 @@ import Redis
 import Storage
 import Sugar
 import Vapor
+import HealthCheck
```
Comment (siemensikkema) [inline Sources/App/Setup/configure.swift]: abc :)

File: Sources/App/Setup/routes.swift [hunk 5]
```diff
@@ -1,5 +1,6 @@
 import AdminPanel
 import Vapor
+import HealthCheck
```
Comment (siemensikkema) [inline Sources/App/Setup/routes.swift:2]: abc 🙂

File: Sources/HealthCheck/Health.swift [hunk 6]
```diff
@@ -0,0 +1,79 @@
+// Health check API
+// see https://inadarei.github.io/rfc-healthcheck/ for more info
+import Vapor
+
+// MARK: Health
+// Structs for generating the Health API response
+public struct Health {
+    // Health status indication options
+    enum Status: String {
```
Comment (siemensikkema) [inline Sources/HealthCheck/Health.swift]: You can make this conform to `Equatable` as well so you don't have to compare on `rawValue`

File: Sources/HealthCheck/Health.swift [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            self.status = status.rawValue
+
+            self.checks = checks.reduce([:], { (result, indicator) in
+                var result = result
+                let key = indicator.key
+                result[key, default: []].append(indicator)
+                return result
+            })
+
+            self.notes = [
```
Comment (siemensikkema) [inline Sources/HealthCheck/Health.swift]: shouldn't this come from the checks you pass in?

File: Sources/HealthCheck/System.swift [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+    func health(on request: Request) -> EventLoopFuture<Response> {
+        self.components.map { $0.healthCheck(on: request) }
+            .flatten(on: request)
+            .flatMap(to: Response.self) { results in
+                var responseStatus: HTTPStatus = .ok
+                var healthStatus: Health.Status = .pass
+
+                results.forEach { check in
+                    if check.status == Health.Status.warn.rawValue {
```
Comment (siemensikkema) [inline Sources/HealthCheck/System.swift]: `if check.status == .warn {`

Comment (siemensikkema) [reply Sources/HealthCheck/System.swift]: with a `for check in checks` you can `break` once you found the first warning. And combine it with the check for failures like so:
```swift
forLoop: for check in checks {
  switch check.status {
  case .warn: healthStatus = .warn
  case .fail: 
    healthStatus = .fail
    responseStatus = .internalServerError
    break forLoop
  default: break
  }
}
```
See https://www.hackingwithswift.com/example-code/language/how-to-break-out-of-multiple-loop-levels-using-labeled-statements

File: Sources/HealthCheck/Health.swift (hunk 6 above)
Comment (rasmusebbesen) [reply Sources/HealthCheck/Health.swift]: Nice :D I have actually been trying to avoid that.

File: Sources/HealthCheck/Health.swift (hunk 7 above)
Comment (rasmusebbesen) [reply Sources/HealthCheck/Health.swift]: I agree

**Review by siemensikkema — APPROVED** [review]
Love it!

**Pull request merged**